OPHD/States/MapViewStateTurn.cpp

```diff
@@ -165,36 +165,57 @@ void MapViewState::updateMorale()
-
-	int food_production = structureManager.getCountInState(Structure::StructureClass::FoodProduction, StructureState::Operational);
-	mCurrentMorale += food_production > 0 ? food_production : -5;
-
-	mCurrentMorale += structureManager.getCountInState(Structure::StructureClass::Commercial, StructureState::Operational);
+	const int birthCount = mPopulation.birthCount();
+	const int parkCount = structureManager.getCountInState(Structure::StructureClass::Park, StructureState::Operational);
+	const int recreationCount = structureManager.getCountInState(Structure::StructureClass::RecreationCenter, StructureState::Operational);
+	const int foodProducingStructures = structureManager.getCountInState(Structure::StructureClass::FoodProduction, StructureState::Operational);
+	const int commercialCount = structureManager.getCountInState(Structure::StructureClass::Commercial, StructureState::Operational);
+	
 
 	// NEGATIVE MORALE EFFECTS
 	// =========================================
-	mCurrentMorale -= mPopulation.deathCount();
-	mCurrentMorale -= structureManager.disabled();
-	mCurrentMorale -= structureManager.destroyed();
-
-	int residentialMoraleHit = mPopulationPanel.capacity() / 100;
-
-	// Ensure that there is always a morale hit if residential capacity is more than 100%.
-	if (mPopulationPanel.capacity() > 100 && residentialMoraleHit < constants::MINIMUM_RESIDENCE_OVERCAPACITY_HIT) { residentialMoraleHit = constants::MINIMUM_RESIDENCE_OVERCAPACITY_HIT; }
+	const int deathCount = mPopulation.deathCount();
+	const int structuresDisabled = structureManager.disabled();
+	const int structuresDestroyed = structureManager.destroyed();
+	
+	const int residentialOverCapacityHit = mPopulation.size() > mResidentialCapacity ? -2 : 0;
 
 	auto& residences = NAS2D::Utility<StructureManager>::get().structureList(Structure::StructureClass::Residence);
+	int bioWasteAccumulation = 0;
 	for (auto residence : residences)
 	{
 		Residence* unit = static_cast<Residence*>(residence);
-		if (unit->wasteOverflow() > 0) { mCurrentMorale -= 2; } /// \fixme magic number
+		if (unit->wasteOverflow() > 0) { ++bioWasteAccumulation; }
 	}
 
-	mCurrentMorale -= residentialMoraleHit;
+	// positive
+	mCurrentMorale += birthCount;
+	mCurrentMorale += parkCount;
+	mCurrentMorale += recreationCount;
+	mCurrentMorale += commercialCount;
+	mCurrentMorale += foodProducingStructures > 0 ? 0 : -5;
+
+	// negative
+	mCurrentMorale -= deathCount;
+	mCurrentMorale -= residentialOverCapacityHit;
+	mCurrentMorale -= bioWasteAccumulation * 2;
+	mCurrentMorale -= structuresDisabled;
+	mCurrentMorale -= structuresDestroyed;
+
 
 	mCurrentMorale = std::clamp(mCurrentMorale, 0, 1000);
+
+	// Provide reason values for PopulationPanel
+	if (birthCount > 0) { mPopulationPanel.addMoraleReason("Births", birthCount); }
+	if (deathCount > 0) { mPopulationPanel.addMoraleReason("Deaths", -deathCount); }
```

Comment:
There is effectively some code duplication here. The morale benefits and penalties from the section above are being duplicated in this section in the second parameter to `addMoraleReason`. If instead the `addMoraleReason` function was responsible for updating the running tally, that could ensure consistency between the two, without code duplication. As it is now, it would be very easy to update only one section with a new morale modifier, or make a typo between the two sections, and so the morale reasons might not match up with actual morale modifiers.

----

Edit: Looks like you figured this out in a later commit.

Additionally, the `if` checks are another area of duplicate code. Effectively the `if` checks are checking for non-zero morale modifiers, and if it's non-zero, to list the text reason along with the modifier value. Instead, the `if` check can be moved into `addMoraleReason`. This of course means a slight update to the `if` condition, since the original code infers a non-zero modifier from a base value, rather than from the morale modifier calculated from it.

To be more concrete, here's an example where the base value `residentialOverCapacityHit` is different from the morale modifier `bioWasteAccumulation * -2`  (wait, why are these different variables?):
```cpp
if (residentialOverCapacityHit > 0) { mPopulationPanel.addMoraleReason("Biowaste Overflowing", bioWasteAccumulation * -2); }
```

Reply:
This would be a mistake that I'm pretty sure I already ironed out in a later commit? Gonna have to double check that. (double checked, it is ironed out in a later commit).